sw/device/tests/crypto/hmac_multistream_functest.c

```diff
@@ -0,0 +1,232 @@
+    case kHmacTestOperationSha512:
+      TRY(otcrypto_hash_update(hash_ctx, msg));
+      break;
+    case kHmacTestOperationHmacSha256:
+      OT_FALLTHROUGH_INTENDED;
+    case kHmacTestOperationHmacSha384:
+      OT_FALLTHROUGH_INTENDED;
+    case kHmacTestOperationHmacSha512:
+      TRY(otcrypto_hmac_update((otcrypto_hmac_context_t *)hash_ctx, msg));
+      break;
+    default:
+      return OTCRYPTO_BAD_ARGS;
+  }
+  return OTCRYPTO_OK;
+}
+
+/**
+ * Finalize the hash/hmac operation with context `hash_ctx`. Further necessary
+ * parameters for the final call is reads from `current_test_vector`.
+ *
+ * @param hash_ctx Corresponding context for given `current_test_vector`.
+ * @param current_test_vector Pointer to the hardcoded test vector.
+ */
+static status_t hmac_finalize(otcrypto_hash_context_t *hash_ctx,
+                              hmac_test_vector_t *current_test_vector) {
+  // The test vectors already have the correct digest sizes hardcoded.
+  size_t digest_len = current_test_vector->digest.len;
+  // Allocate the buffer for the maximum digest size (which comes from SHA-512).
+  uint32_t act_tag[kSha512DigestWords];
+  otcrypto_word32_buf_t tag_buf = {
+      .data = act_tag,
+      .len = digest_len,
+  };
+  otcrypto_hash_digest_t hash_digest = {
+      // .mode is to be determined below in switch-case block.
+      .data = act_tag,
+      .len = digest_len,
+  };
+  LOG_INFO("Invoking final call for %s.",
+           current_test_vector->vector_identifier);
+  switch (current_test_vector->test_operation) {
+    case kHmacTestOperationSha256:
+      OT_FALLTHROUGH_INTENDED;
+    case kHmacTestOperationSha384:
+      OT_FALLTHROUGH_INTENDED;
+    case kHmacTestOperationSha512:
+      TRY(get_hash_mode(current_test_vector, &hash_digest.mode));
+      TRY(otcrypto_hash_final(hash_ctx, hash_digest));
+      break;
+    case kHmacTestOperationHmacSha256:
+      OT_FALLTHROUGH_INTENDED;
+    case kHmacTestOperationHmacSha384:
+      OT_FALLTHROUGH_INTENDED;
+    case kHmacTestOperationHmacSha512:
+      TRY(otcrypto_hmac_final((otcrypto_hmac_context_t *)hash_ctx, tag_buf));
+      break;
+    default:
+      return OTCRYPTO_BAD_ARGS;
+  }
+  LOG_INFO("Comparing result for %s.", current_test_vector->vector_identifier);
+  TRY_CHECK_ARRAYS_EQ(act_tag, current_test_vector->digest.data, digest_len);
+  return OTCRYPTO_OK;
+}
+
+/**
+ * Run all vectors specified in `kHmacTestVectors` in parallel. Namely:
+ * i) Each stream is instantiated (by init call).
+ * ii) Each stream receives the first half its message.
+ * iii) Each stream receives the second half of its message.
+ * iv) Each stream is concluded with a final call and the digest result is
+ * compared with the expected valud.
+ */
+static status_t run_test(void) {
+  // hash_contexts are also used as hmac_contexts, since their size is equal.
+  otcrypto_hash_context_t hash_contexts[ARRAYSIZE(kHmacTestVectors)];
+  for (size_t i = 0; i < ARRAYSIZE(kHmacTestVectors); i++) {
```

Comment:
Would there be any value from randomizing all these loops from a DV coverage perspective? If so we can talk about how to do that in a follow up PR.

Reply:
Yes randomizing the order here would be great. I was not sure if there are available utils regarding randomness and where to get this randomness. I see `sw/device/lib/base/random_order.h` but the implementation looks like a place holder.

At the moment, this test does not configure EDN to save some time, but we can reenable it if it helps with the randomness.

Reply:
An alternative would be to make some bazel changes so that the hardcoded testvector file `sw/device/tests/crypto/testvectors/hmac_testvectors_random.hjson` is re-generated randomly with the following:
```
./sw/device/tests/crypto/hmac_gen_random_testvectors.py 10 sw/device/tests/crypto/testvectors/hmac_testvectors_random.hjson
```
The generation script does not take seed at the moment, but I can add it. It would put less pressure on complication to TLT and we can avoid configuring EDN I think.

In any case, I think we can merge this once CI is clean and discuss how to improve the test. It is also possible to run this test with more vectors on FPGA/nightly.